workflows/src/main/kotlin/com/template/flows/CreateAppointmentRequest.kt

```diff
@@ -21,52 +21,67 @@ import net.corda.core.transactions.TransactionBuilder
 import com.template.states.AvailableAppointmentDate
 import net.corda.core.contracts.StateAndRef
 import net.corda.core.identity.AbstractParty
+import net.corda.core.node.services.StatesNotAvailableException
 
 // *********
 // * Flows *
 // *********
 @InitiatingFlow
 @StartableByRPC
 class CreateAppointmentRequest(private val doctor: Party,
-                               private val date: String,
-                               private val appointmentDate: StateAndRef<AvailableAppointmentDate>) : FlowLogic<SignedTransaction>() {
+                               private val date: String) : FlowLogic<SignedTransaction>() {
     override val progressTracker = ProgressTracker()
 
     @Suspendable
     override fun call(): SignedTransaction {
+        lateinit var appointmentDate: StateAndRef<AvailableAppointmentDate>
 
-        val patient = ourIdentity
+        val vaultAppointmentDates = serviceHub.vaultService.queryBy(AvailableAppointmentDate::class.java).states
```

Comment:
You can query by date and the loop below won't be required, especially that AvailableAppointmentDate is already a QueryableState.